### ISSUE TITLE
Serialization issues

### ISSUE THREAD

**Issue by vyazelenko** [real-logic/agrona#223]
I've found several issues with how `Serializable` is used in the project:

- No classes that are `Serializable` declare `serialVersionUID` fields (e.g. [AsciiNumberFormatException](https://github.com/real-logic/agrona/blob/cb97cda84c4fe0f61a71e71ff73e86971d84c408/agrona/src/main/java/org/agrona/AsciiNumberFormatException.java), [Int2ObjectHashMap](https://github.com/real-logic/agrona/blob/cb97cda84c4fe0f61a71e71ff73e86971d84c408/agrona/src/main/java/org/agrona/collections/Int2ObjectHashMap.java) etc.).

   The classes should declare `serialVersionUID` field but it must be compatible with the old code, i.e. the value of the `serialVersionUID` field must match values at least from the latest released versions of Agrona on JDK 8 to avoid breaking existing code. This needs to also work for generated classes.

- Some serializable classes contain non-serializable fields:
  - Externally assigned fields - for example [IntLruCache](https://github.com/real-logic/agrona/blob/cb97cda84c4fe0f61a71e71ff73e86971d84c408/agrona/src/main/java/org/agrona/collections/IntLruCache.java) takes two parameters one `java.util.function.IntFunction` and another `java.util.function.Consumer` neither of which are serializable by default and assigns them to the non-transient fields.

     The easiest solution would be to mark the corresponding fields `transient` if it would not break the internal logic of the class. Otherwise the solution is require provided functions to be serializable. For example in case of [IntLruCache](https://github.com/real-logic/agrona/blob/cb97cda84c4fe0f61a71e71ff73e86971d84c408/agrona/src/main/java/org/agrona/collections/IntLruCache.java) the constructor signature will change to this:
     ```java
          public <FACTORY extends IntFunction<E> & Serializable, CONSUMER extends Consumer<E> & Serializable> IntLruCache(
             @DoNotSub final int capacity,
             final FACTORY factory,
             final CONSUMER closer)
          {
             ...
          }
     ```
     Or maybe simple `instanceof Serializable` check when creating object would suffice.
  - Internal fields - for example [Int2ObjectCache](https://github.com/real-logic/agrona/blob/cb97cda84c4fe0f61a71e71ff73e86971d84c408/agrona/src/main/java/org/agrona/collections/Int2ObjectCache.java) caches `KeySet`, `ValueCollection` and `EntrySet` none of which are serializable. In this case we can either mark fields as transient or make the corresponding classes serializable.

**Comment by mjpt777** [comment]
I was against added Serializable but some people convince me it was just for testing. We should not need to be backwards compatible between versions with a `serialVersionUID` change.

**Comment by vyazelenko** [comment]
@mjpt777 I guess we keep non-serializable lambdas as is, i.e. they render the corresponding collections non-serializable despite those implementing `Serializable` interface. 

Possible solutions:
- Drop `Serializable` from the collections that accept lambda as the constructor argument.
- Add run-time check to ensure that provided lambda is serializable. **_Will break all existing code!_**
- Add compile-time check check to ensure that provided lambda is serializable. **_Will break all existing code!_**
- Add run-time check when serializing to ensure that use gets a nicer exception.
- Do nothing - let it burn when someone tries to serialize.

**Comment by mjpt777** [comment]
If things cannot be seralized then we should remove it. Which classes are impacted?

**Comment by vyazelenko** [comment]
@mjpt777 Updated #224 to use `serialVersionUID` instead of `@SuppressWarnings("serial")`.

**Comment by vyazelenko** [comment]
Impacted classes:
- `org.agrona.collections.Int2ObjectCache`
- `org.agrona.collections.IntLruCache`
- `org.agrona.collections.UnmodifiableCollectionView`

And the generated classes from them.

**Comment by mjpt777** [comment]
We can remove `Serializable` from them.

**Comment by vyazelenko** [comment]
Ok, will do in my PR/

**Comment by vyazelenko** [comment]
The rest of this issues should be fixed with #224.

**Comment by guidomedina** [comment]
I'm too late to this discussion but the reason there were `Serializable` was because of this issue https://github.com/real-logic/agrona/issues/110
